### PR TITLE
PBD-137, Weighted voting

### PR DESCRIPTION
**Pull request by CryptoCubik**

### CODE REVIEW

**Comment by CryptoCubik** [comment]
There is an issue where proposals are created locally after calling `proposal post` even if we don't have enough PART to send the message.

**Comment by CryptoCubik** [comment]
Originally I think we were planning not to create proposal results until votes were received; no we create them immediately upon proposal creation.
This caused issues with votes being counted.
That's been fixed now, but now we can vote more than once for some reason.

**Comment by CryptoCubik** [comment]
Currently there's a vulnerability where you can vote, move your PART to a new profile address, then vote again, potentially infinite times, rigging an item vote.
We should probably patch that before the first beta

**Comment by CryptoCubik** [comment]
Currently item votes are ended after a certain number of votes.
As such if we just recalculate the vote weights every time we receive a new vote, it should fix the problem completely.

This does, however, open the possibility of votes with zero weight being counted towards the number of votes (currently we ignore/don't-send votes that have no weight); to remedy this we should exclude zero weighted votes from the vote count before calculating if a vote should end.

**Comment by rynomster** [comment]
Yes, we need to recalculate everytime a new vote comes in, and if a vote
has no weight, its not a valid vote.

>

**Comment by CryptoCubik** [comment]
Votes now cast one per wallet address, not profile

File: src/api/models/Vote.ts [hunk 1]
```diff
@@ -47,6 +47,26 @@ export class Vote extends Bookshelf.Model<Vote> {
         }
     }
 
+    public static async fetchAllFromMeByProposalId(proposalId: number, withRelated: boolean = true): Promise<Collection<Vote>> {
```
Comment (xludx) [inline src/api/models/Vote.ts]: what does this mean? "FromMe"? the functionality seems to implement fetchAllByProposal.

File: src/api/services/ProposalActionService.ts [hunk 2]
```diff
@@ -90,20 +98,83 @@ export class ProposalActionService {
...
+            const paidMessage = false;
+
+            /*
+             * Vote once per profile
+             */
+            if (!estimateFee) {
+                let proposal;
+                try {
+                    proposal = await this.processProposal(proposalCreateRequest);
+                } catch (ex) {
```
Comment (xludx) [inline src/api/services/ProposalActionService.ts]: remove the catch, if processing fails then we shouldn't just continue

File: src/api/services/ProposalActionService.ts [hunk 3]
```diff
@@ -90,20 +98,83 @@ export class ProposalActionService {
...
+
+                // Get the default market.
+                //     TODO: Might want to let users specify this later.
+                const marketModel = await this.marketService.getDefault();
+                if (!marketModel) {
+                    throw new MessageException(`Default market doesn't exist!`);
+                }
+                const market: resources.Market = marketModel.toJSON();
+
+                // For each profile, vote.
```
Comment (xludx) [inline src/api/services/ProposalActionService.ts]: address, not profile

File: src/api/services/ProposalActionService.ts [hunk 4]
```diff
@@ -216,64 +312,68 @@ export class ProposalActionService {
...
+            // also create the FlaggedItem
+            const flaggedItem = await this.createFlaggedItemForProposal(createdProposal);
+            // this.log.debug('processItemVoteProposal(), flaggedItem:', JSON.stringify(flaggedItem, null, 2));
+
+            createdProposalModel = await this.proposalService.findOne(createdProposal.id);
+            proposal = createdProposalModel.toJSON();
         }
+        this.log.debug('processProposal. proposal = ' + JSON.stringify(proposal, null, 2));
+        this.log.debug('processProposal, Doing findOneByProposalHash');
+        let proposalResult: any = this.proposalResultService.findOneByProposalHash(proposal.hash);
```
Comment (xludx) [inline src/api/services/ProposalActionService.ts]: will throw if not found...

File: src/api/services/VoteService.ts [hunk 5]
```diff
@@ -12,18 +12,33 @@ import { VoteRepository } from '../repositories/VoteRepository';
...
         @inject(Types.Repository) @named(Targets.Repository.VoteRepository) public voteRepo: VoteRepository,
+        @inject(Types.Service) @named(Targets.Service.CoreRpcService) public coreRpcService: CoreRpcService,
         @inject(Types.Core) @named(Core.Logger) public Logger: typeof LoggerType
     ) {
         this.log = new Logger(__filename);
     }
 
+    public async getVoteWeight(voter: string): Promise<number> {
+        let voteWeight;
+        try {
```
Comment (xludx) [inline src/api/services/VoteService.ts]: remove try catch

File: src/api/services/VoteService.ts [hunk 6]
```diff
@@ -46,6 +61,15 @@ export class VoteService {
         return vote;
     }
 
+    public async findAllFromMeByProposalId(proposalId: number, withRelated: boolean = true): Promise<Bookshelf.Collection<Vote>> {
```
Comment (xludx) [inline src/api/services/VoteService.ts]: findAllByProposalId, this has nothing to do with "me".

**Comment by xludx** [comment]
https://github.com/particl/particl-market/pull/359

**Pull request closed without merging**